testutil/network.go

```diff
@@ -0,0 +1,358 @@
+			WithKeybase(kb)
+
+		signedTx, err := txBldr.SignStdTx(nodeDirName, tx, false)
+		require.NoError(t, err)
+
+		txBz, err := cdc.MarshalJSON(signedTx)
+		require.NoError(t, err)
+		require.NoError(t, writeFile(fmt.Sprintf("%v.json", nodeDirName), gentxsDir, txBz))
+
+		srvconfig.WriteConfigFile(filepath.Join(nodeDir, "config/app.toml"), appCfg)
+
+		network.Validators[i] = &Validator{
+			AppConfig:  appCfg,
+			Ctx:        ctx,
+			Dir:        filepath.Join(network.BaseDir, nodeDirName),
+			NodeID:     nodeID,
+			PubKey:     pubKey,
+			Moniker:    nodeDirName,
+			RPCAddress: tmCfg.RPC.ListenAddress,
+			P2PAddress: tmCfg.P2P.ListenAddress,
+			APIAddress: apiAddr,
+			Address:    addr,
+			ValAddress: sdk.ValAddress(addr),
+		}
+	}
+
+	require.NoError(t, initGenFiles(cfg, genAccounts, genBalances, genFiles))
+	require.NoError(t, collectGenFiles(cfg, network.Validators, network.BaseDir))
+
+	t.Log("starting test network...")
+	for _, v := range network.Validators {
+		require.NoError(t, startInProcess(cfg, v))
+	}
+
+	t.Log("started test network")
+
+	// Ensure we cleanup incase any test was abruptly halted (e.g. SIGINT) as any
+	// defer in a test would not be called.
+	server.TrapSignal(network.Cleanup)
+
+	return network
+}
+
+// WaitForHeight performs a blocking check where it waits for a block to be
+// committed after a given block. If that height is not reached within a timeout,
+// an error is returned. Regardless, the latest height queried is returned.
+func (n *Network) WaitForHeight(h int64) (int64, error) {
+	return n.WaitForHeightWithTimeout(h, 10*time.Second)
+}
+
+// WaitForHeightWithTimeout is the same as WaitForHeight except the caller can
+// provide a custom timeout.
+func (n *Network) WaitForHeightWithTimeout(h int64, t time.Duration) (int64, error) {
+	ticker := time.NewTicker(time.Second)
+	timeout := time.After(t)
+
+	if len(n.Validators) == 0 {
+		return 0, errors.New("no validators available")
+	}
+
+	var latestHeight int64
+	val := n.Validators[0]
+
+	for {
+		select {
+		case <-timeout:
+			ticker.Stop()
+			return latestHeight, errors.New("timeout exceeded waiting for block")
+		case <-ticker.C:
+			status, err := val.RPCClient.Status()
+			if err == nil && status != nil {
+				latestHeight = status.SyncInfo.LatestBlockHeight
+				if latestHeight >= h {
+					return latestHeight, nil
+				}
+			}
+		}
+	}
+}
+
+// Cleanup removes the root testing (temporary) directory and stops both the
+// Tendermint and API services. It allows other callers to create and start
+// test networks. This method must be called when a test is finished, typically
+// in a defer.
+func (n *Network) Cleanup() {
+	defer func() {
+		lock.Unlock()
+		n.T.Log("released test network lock")
+	}()
+
+	n.T.Log("cleaning up test network...")
+
+	for _, v := range n.Validators {
+		if v.tmNode != nil && v.tmNode.IsRunning() {
+			_ = v.tmNode.Stop()
+		}
+
+		if v.api != nil {
+			_ = v.api.Close()
+		}
+	}
+
+	_ = os.RemoveAll(n.BaseDir)
```

Comment:
We should definitely have some way to cleanup all the test directories. But wouldn't some of these files be useful for debugging?

If they are always deleted on cleanup, I can't look through any of them to figure out why something failed

Reply:
Good idea. I'll add a config (defaults to removing) and print the path in case its useful for post-test resolution.

Reply:
Pushed.